src/fabric/src/fabric_drop_seq.erl

```diff
@@ -0,0 +1,1020 @@
+            end,
+            ShardSyncHistory1 = maps:merge(
+                maps:groups_from_list(KeyFun, ValueFun, History), ShardSyncHistory0
+            ),
+            {ok, {PeerCheckpoints, ShardSyncHistory1}}
+    end;
+gather_drop_seq_info_fun(#doc{}, Acc) ->
+    %% ignored
+    {ok, Acc}.
+
+gather_drop_seq_info_cb({rexi_DOWN, _, _, _}, _Worker, {Acc, Count}) ->
+    {ok, {Acc, Count - 1}};
+gather_drop_seq_info_cb({rexi_EXIT, _Reason}, _Worker, {Acc, Count}) ->
+    {ok, {Acc, Count - 1}};
+gather_drop_seq_info_cb({ok, Info}, Worker, {Acc, Count}) ->
+    MergedInfo = merge_info(Worker, Info, Acc),
+    if
+        Count == 0 ->
+            {stop, MergedInfo};
+        true ->
+            {ok, {MergedInfo, Count - 1}}
+    end;
+gather_drop_seq_info_cb(_Error, _Worker, {Acc, Count}) ->
+    {ok, {Acc, Count - 1}}.
+
+merge_info(#shard{} = Shard, Info, Acc) ->
+    #{
+        uuid_map =>
+            maps:put(
+                {Shard#shard.range, Shard#shard.node}, maps:get(uuid, Info), maps:get(uuid_map, Acc)
+            ),
+        peer_checkpoints => maps:merge_with(
+            fun merge_peers/3,
+            maps:get(peer_checkpoints, Info),
+            maps:get(peer_checkpoints, Acc)
+        ),
+        shard_sync_history => maps:merge(
+            maps:get(shard_sync_history, Info), maps:get(shard_sync_history, Acc)
+        )
+    }.
+
+merge_peers(_Key, {Uuid1, Val1}, {Uuid2, Val2}) when
+    is_binary(Uuid1), is_binary(Uuid2), is_integer(Val1), is_integer(Val2)
+->
+    ?assert(uuids_match([Uuid1, Uuid2]), "UUIDs belong to different shard files"),
+    {Uuid1, min(Val1, Val2)}.
+
+uuids_match(Uuids) when is_list(Uuids) ->
+    PrefixLen = lists:min([byte_size(Uuid) || Uuid <- Uuids]),
+    binary:longest_common_prefix(Uuids) == PrefixLen.
+
+decode_seq(OpaqueSeq) ->
+    Decoded = fabric_view_changes:decode_seq(OpaqueSeq),
+    lists:foldl(
+        fun
+            ({_Node, [S, E], {Seq, Uuid, Node}}, Acc) when
+                is_integer(S),
+                is_integer(E),
+                is_integer(Seq),
+                S >= 0,
+                E > S,
+                Seq >= 0,
+                is_binary(Uuid),
+                is_atom(Node)
+            ->
+                Acc#{{[S, E], Node} => {Uuid, Seq}};
+            (_Else, Acc) ->
+                Acc
+        end,
+        #{},
+        Decoded
+    ).
+
+latest_shard_sync_checkpoints(ShardSyncHistory) ->
+    maps:fold(
+        fun({R, SN, _TN}, History, Acc) ->
+            {SU, SS, _TU, _TS} = hd(History),
+            maps:merge_with(fun merge_peers/3, #{{R, SN} => {SU, SS}}, Acc)
+        end,
+        #{},
+        ShardSyncHistory
+    ).
+
+%% A shard may have been split since a peer saw it.
+-spec substitute_splits([#shard{}], uuid_map(), peer_checkpoints()) -> peer_checkpoints().
+substitute_splits(Shards, UuidMap, PeerCheckpoints) ->
+    maps:fold(
+        fun({[PS, PE], Node}, {Uuid, Seq}, Acc) ->
+            ShardsInRange = [
```

Comment:
We could re-use this functionality from mem3.

```erlang
%% mem3.erl

filter_shards_by_range(Range, Shards) ->
    mem3_shards:filter_shards_by_range(Range, Shards)

%% mem3_shards.erl

-export([filter_shards_by_range/2]).
```

Reply:
hm, I don't think it's the same calculation. for splits I want to find the ranges that are subsets of the peer checkpoints subset, it wouldn't be enough just for them to overlap.